### ISSUE TITLE
No warning for obsolete constructor when using generic `new()` constraint

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#61009]
Inspired from #61005

The following code compiles successfully:

```csharp
using System;

C.M<ThisIsObsolete>();

public class ThisIsObsolete
{
    [Obsolete("This is obsolete", error: true)]
    public ThisIsObsolete() { }
}

public static class C
{
    public static T M<T>() where T : new()
    {
        return new T();
    }
}
```

Is that expected? If not, should this be added to a warning wave? should a breaking change be taken?

**Comment by jcouv** [comment]
@333fred Reporting the obsolete diagnostic seems sensible. The [spec](https://github.com/dotnet/csharpstandard/blob/670fc242ff069941defef1df89627edfdd5dfd1c/standard/attributes.md#2154-the-obsolete-attribute) doesn't go into much details on obsolete.
What do you think?
Let's check with compat council. Seems we have 3 options: 1. report the missing diagnostic, 2. report it but downgraded to a warning wave, 3. status quo.

**Comment by CyrusNajmabadi** [comment]
It's definitely a bit of a head scratcher.  The code is just saying "you need to have this method" not that it will be called.  It would be somewhat similar to doing this:

```c#
interface IFoo
{
    void Foo();
}

class Foo
{
    [Obsolete]
    public void Foo() { }
}

// ...

void M<T>(T t) where T : IFoo { }

// ...

M(new Foo());
```

Should this give an error/warning?  M says it needs an IFoo which means it *may* call `IFoo.Foo`, which on this type would be an error.  But it doesn't mean it *will* call it.  Similarly, the `new()` constraint means it may new up the object, not that it will.  So there's a bit of a disconnect.  

However, maybe it's much more likely in the constructor-constraint that it is likely to be called, and thus that obsolete constructor will be used and will be a problem.

I'm honestly not sure...

**Comment by jcouv** [comment]
I find those two situations (`new()` vs. interface constraint) not that similar.
If the code didn't rely on the parameterless constructor, then the `new()` constraint could just be omitted. It serves no other purpose than to indicate "I intend to instantiate this type".
On the other hand, I could see scenarios where you constrain on an interface, but don't use certain (or any) members of the interface.

**Comment by 333fred** [comment]
We've had locations were we didn't report obsolete diagnostics before: https://github.com/dotnet/roslyn/issues/47308. Let's start by checking to see what the native compiler did in this scenario. If they reported an error, I think we'd be safe to just report. If it didn't we may want to consider only checking with a warning wave turned on. Either way, followup with compat council is needed.

**Comment by alrz** [comment]
Previous discussion on the subject: https://github.com/dotnet/roslyn/issues/19430

**Comment by 333fred** [comment]
Closing as a duplicate of #19430
